org.testeditor.web.backend.persistence/src/main/java/org/testeditor/web/backend/persistence/DocumentProvider.xtend

```diff
@@ -47,39 +58,72 @@ class DocumentProvider {
 		return folder.mkdirs
 	}
 
-	/**
-	 * @return true when the file has been created
-	 */
-	def boolean createOrUpdate(String resourcePath, String content) {
-		val file = getWorkspaceFile(resourcePath)
-		var created = false
-		if (!file.exists) {
-			created = create(file)
-		}
-		file.write(content)
-		return created
-	}
-
 	def String load(String resourcePath) {
 		val file = getWorkspaceFile(resourcePath)
-		if (!regardAsBinary(resourcePath)) {
-			return file.read
+		pull
+
+		if (file.exists) {
+			if (!regardAsBinary(resourcePath)) {
+				return file.read
+			} else {
+				throw new IllegalStateException('''File "«file.name»" appears to be binary and cannot be loaded as text.''')
+			}
 		} else {
-			throw new IllegalStateException('''File "«file.name»" appears to be binary and cannot be loaded as text.''')
+			throw new FileNotFoundException('''The file '«resourcePath»' does not exist. It may have been concurrently deleted.''')
 		}
 	}
 
 	def void save(String resourcePath, String content) {
-		val file = getWorkspaceFile(resourcePath)
-		file.write(content)
+		resourcePath.getWorkspaceFile => [
+			Files.asCharSink(it, UTF_8).write(content)
+			commit('''update file: «it.name»''')
+		]
+
+		repoSync[onConflict|onConflict.resetToRemoteAndCreateBackup(resourcePath, content)]
+	}
+
+	private def String getConflictMessage(StageState conflictState, String resourcePath) {
+		return switch (conflictState) {
+			case BOTH_MODIFIED: '''The file '«resourcePath»' could not be saved due to concurrent modifications.'''
+			case DELETED_BY_THEM: '''The file '«resourcePath»' could not be saved as it was concurrently being deleted.'''
+			case BOTH_ADDED: '''The file '«resourcePath»' already exists.'''
+			case DELETED_BY_US: '''The file '«resourcePath»' could not be deleted as it was concurrently modified.'''
+			case BOTH_DELETED,
+			case ADDED_BY_THEM,
+			case ADDED_BY_US: '''Don't know how to handle conflict: «conflictState.toString»'''
+		}
+	}
+
+	private def String appendBackupNote(String conflictMessage, String resourcePath) {
+		return conflictMessage + ''' Local changes were instead backed up to '«resourcePath».local-backup'.'''
+	}
+
+	private def void resetToRemoteState() {
+		git.reset => [
+			ref = 'origin/master'
```

Comment:
doesn't this reset to head of the master branch, what about other branches? are we currently working on master branch only? we should then put this into a configuration, probably.

Reply:
resetting to `@{upstream}` should solve that issue without the need to make this configurable.